README.md
# Slim-SessionDatabase

## Features

- This is fork of [stefangabos/Zebra_Session](https://github.com/stefangabos/Zebra_Session).

## Changes from stefangabos/Zebra_Session

* Helper function is removed.(get_settings(), regenerate_id(), get_active_sessions(), stop(), ...)
* Not support flashdata.
* Use tab indentation.
* Not support composer yet.
* Support database driver. (Mysqli, PDO, PDO-Mysql) 

## Requirements

* PHP > 5.4.0 (namespace, SessionHandlerInterface)
* Mysql > 4.1.22

## Installation

Library is not registered in packagist yet.

> composer.json

```json
{
    "repositories": [
        { "type": "vcs", "url": "https://github.com/steamegg/Slim-SessionDatabase.git"}
    ],
    "require": {
        "steamegg/Slim-SessionDatabase": "dev-develop"
    }
}
```

## Install MySQL table

> install/session_data.sql

```sql
CREATE TABLE `session_data` (
  `session_id` varchar(32) NOT NULL default '',
  `hash` varchar(32) NOT NULL default '',
  `session_data` blob NOT NULL,
  `session_expire` int(11) NOT NULL default '0',
  PRIMARY KEY  (`session_id`)
) DEFAULT CHARSET=utf8;
```

## How to use

> Create database connection using Mysqli
```php
use steamegg\Slim\SessionDatabase\Connection\MysqliConnection;
use steamegg\Slim\SessionDatabase\SessionConfig;
use steamegg\Slim\SessionDatabase\SessionDbHandler;

$connection = new MysqliConnection(mysqli_connect("localhost","dbuser","password","test"));
$config = new SessionConfig(
			"SECURITY", 
			$_SERVER["HTTP_USER_AGENT"],
			ini_get("session.gc_maxlifetime"),
			ini_get("session.gc_probability"),
			ini_get("session.gc_divisor")
			);
new SessionDbHandler($connection, $config);
```

> Create database connection using PDO-Mysql
```php
use steamegg\Slim\SessionDatabase\Connection\PdoMysqlConnection;
use steamegg\Slim\SessionDatabase\SessionConfig;
use steamegg\Slim\SessionDatabase\SessionDbHandler;

$connection = new PdoMysqlConnection(new \PDO("mysql:dbname=test;host=localhost", "dbuser", "password"));
$config = new SessionConfig(
			"SECURITY", 
			$_SERVER["HTTP_USER_AGENT"],
			ini_get("session.gc_maxlifetime"),
			ini_get("session.gc_probability"),
			ini_get("session.gc_divisor")
			);
new SessionDbHandler($connection, $config);
```

## Fingerprint
> If you want to use user-agent only
```php
$config = new SessionConfig(
			"SECURITY", 
			$_SERVER["HTTP_USER_AGENT"],
			ini_get("session.gc_maxlifetime"),
			ini_get("session.gc_probability"),
			ini_get("session.gc_divisor")
			);
```

> If you want to use user-agent with ip
```php
$config = new SessionConfig(
			"SECURITY", 
			$_SERVER["HTTP_USER_AGENT"].$_SERVER["REMOTE_ADDR"],
			ini_get("session.gc_maxlifetime"),
			ini_get("session.gc_probability"),
			ini_get("session.gc_divisor")
			);
```

CHANGELOG.md
## 0.9 (2018-02-14)

- Helper function is removed.(get_settings(), regenerate_id(), get_active_sessions(), stop(), ...)
- Not support flashdata.
- Use tab indentation.
- Support database driver. (Mysqli, PDO, PDO-Mysql)

## 0.1 (2018-02-12)

- initial fork release


tests/Test/TestCase.php
<?php
namespace steamegg\Slim\SessionDatabase\Test;

class TestCase extends \PHPUnit_Framework_TestCase {
}


tests/Test/SessionConfigTest.php
<?php
namespace steamegg\Slim\SessionDatabase\Test;

use steamegg\Slim\SessionDatabase\SessionConfig;

class SessionConfigTest extends TestCase {
	
	function testSecurityCode(){
		$config = new SessionConfig("code1", "fingerprint", 111, 222, 333);
		$this->assertAttributeEquals("code1", "security_code", $config);
	}
	
	function testFingerprint(){
		$config = new SessionConfig("code1", "fingerprint", 111, 222, 333);
		$this->assertAttributeEquals("fingerprint", "fingerprint", $config);
		
		$config = new SessionConfig("code1", "fingerprint+ip", 111, 222, 333);
		$this->assertAttributeEquals("fingerprint+ip", "fingerprint", $config);
	}
	
	function testSessionLifetimeShouldNumeric(){
		$this->setExpectedException("PHPUnit_Framework_Error", "session_lifetime is not numeric");
		$config = new SessionConfig("code1", "fingerprint", "notNumeric", 222, 333);
	}
	
	function testSessionLifetime(){
		$config = new SessionConfig("code1", "fingerprint", 111, 222, 333);
		$this->assertEquals(111, $config->getSessionLifetime());
	}
	
	function testGcProbabilityShouldNumeric(){
		$this->setExpectedException("PHPUnit_Framework_Error", "gc_probability is not numeric");
		$config = new SessionConfig("code1", "fingerprint", 111, "notNumeric", 333);
	}
	
	function testGcDivisorShouldNumeric(){
		$this->setExpectedException("PHPUnit_Framework_Error", "gc_divisor is not numeric");
		$config = new SessionConfig("code1", "fingerprint", 111, 222, "notNumeric");
	}
	
	function testGc(){
		$config = new SessionConfig("code1", "fingerprint", 111, 222, 333);
		$this->assertEquals(222, $config->getGcProbability());
		$this->assertEquals(333, $config->getGcDivisor());
	}
	
	function testTable(){
		$config = new SessionConfig("code1", "fingerprint", 111, 222, 333);
		$this->assertEquals("session_data", $config->getTable());
		
		$config = new SessionConfig("code1", "fingerprint", 111, 222, 333, "table1");
		$this->assertEquals("table1", $config->getTable());
	}
	
	function testHash(){
		$config = new SessionConfig("code1", "fingerprint", 111, 222, 333);
		$hash = md5(sprintf("%s%s", "code1", "fingerprint"));
		$this->assertEquals($hash, $config->calculateHash());
		
		$hash = md5(sprintf("%s%s", "code1", "fingerprint+ip"));
		$this->assertNotEquals($hash, $config->calculateHash());
	}
}


src/SessionConfig.php
<?php
namespace steamegg\Slim\SessionDatabase;

class SessionConfig {
	private $security_code;
	private $fingerprint;
	private $session_lifetime;
	private $gc_probability;
	private $gc_divisor;
	private $table;
	
	function __construct(
		$security_code,
		$fingerprint,
		$session_lifetime,
		$gc_probability,
		$gc_divisor,
		$table = 'session_data'){
		
		$this->security_code = $security_code;
		$this->fingerprint = $fingerprint;
		
		if(!is_numeric($session_lifetime))
			trigger_error("session_lifetime is not numeric", E_USER_ERROR);
		$this->session_lifetime = $session_lifetime;
		
		if(!is_numeric($gc_probability))
			trigger_error("gc_probability is not numeric", E_USER_ERROR);
		$this->gc_probability = $gc_probability;
		
		if(!is_numeric($gc_divisor))
			trigger_error("gc_divisor is not numeric", E_USER_ERROR);
		$this->gc_divisor = $gc_divisor;
		
		$this->table = $table;
	}
	
	function getSessionLifetime(){
		return $this->session_lifetime;
	}
	
	function getGcProbability(){
		return $this->gc_probability;
	}
	
	function getGcDivisor(){
		return $this->gc_divisor;
	}
	
	function getTable(){
		return $this->table;
	}
	
	function calculateHash(){
		$hash = sprintf("%s%s", $this->security_code, $this->fingerprint);
		return md5($hash);
	}
}


src/Zebra_Session.php
<?php
namespace steamegg\Slim\SessionMysql;

use steamegg\Slim\SessionMysql\Connection\IConnection;

/**
 *  A PHP library acting as a drop-in replacement for PHP's default session handler, but instead of storing session data
 *  in flat files it stores them in a MySQL database, providing better performance as well as better security and
 *  protection against session fixation and session hijacking.
 *
 *  Read more {@link https://github.com/stefangabos/Zebra_Session/ here}
 *
 *  @author	 Stefan Gabos <contact@stefangabos.ro>
 *  @version	2.1.8 (last revision: May 20, 2017)
 *  @copyright  (c) 2006 - 2017 Stefan Gabos
 *  @license	http://www.gnu.org/licenses/lgpl-3.0.txt GNU LESSER GENERAL PUBLIC LICENSE
 *  @package	Zebra_Session
 */

class Zebra_Session implements \SessionHandlerInterface{

	/**
	 * @var IConnection
	 */
	private $connection;
	
	/**
	 * @var SessionConfig
	 */
	private $config;
	
	/**
	 * the maximum amount of time (in seconds) for which a process can lock the session
	 * @var int
	 */
	private $lock_timeout;
	
	/**
	 * @var string
	 */
	private $lockName;
	
	public function __construct(IConnection &$connection, SessionConfig $config, $lock_timeout = 60) {
		
		if(!$connection->ping())
			trigger_error("Connection ping failed", E_USER_ERROR);

		$this->connection = $connection;
		$this->config = $config;
		$this->lock_timeout = $lock_timeout;
		
		$this->setIni($this->config);

		session_set_save_handler(
			array(&$this, 'open'),
			array(&$this, 'close'),
			array(&$this, 'read'),
			array(&$this, 'write'),
			array(&$this, 'destroy'),
			array(&$this, 'gc')
		);

		session_start();
	}
	
	protected function setIni(SessionConfig $config){
		ini_set('session.gc_maxlifetime', $config->getSessionLifetime());
		ini_set('session.gc_probability', $config->getGcProbability());
		ini_set('session.gc_divisor', $config->getGcDivisor());
		
		// make sure session cookies never expire so that session lifetime
		// will depend only on the value of $session_lifetime
		ini_set('session.cookie_lifetime', 0);
		
		// tell the browser not to expose the cookie to client side scripting
		// this makes it harder for an attacker to hijack the session ID
		ini_set('session.cookie_httponly', 1);
		
		// make sure that PHP only uses cookies for sessions and disallow session ID passing as a GET parameter
		ini_set('session.use_only_cookies', 1);
	}

	function close() {
		if($this->connection->releaseLock($this->lockName))
			return TRUE;
	}

	function destroy($session_id) {
		$sql = sprintf("DELETE FROM %s WHERE session_id = %s", 
			$this->config->getTable(), 
			$this->connection->quote($session_id));
		$this->connection->query($sql);

		return $this->connection->affectedRows() >= 0 ? TRUE : FALSE;
	}

	function gc($maxlifetime) {
		$sql = sprintf("DELETE FROM %s WHERE session_expire < %s", 
			$this->config->getTable(), 
			$this->connection->quote(time()));
		$this->connection->query($sql);
	}

	function open($save_path, $name){
		return TRUE;
	}

	function read($session_id) {
		$this->lockName = sprintf("session_%s", $session_id);

		if( !$this->connection->getLock($this->lockName, $this->lock_timeout) )
			die("Could not obtain session lock");

		$sql = sprintf("SELECT session_data FROM %s WHERE session_id = %s AND session_expire > %s AND hash = %s LIMIT 1", 
			$this->config->getTable(), 
			$this->connection->quote($session_id), 
			$this->connection->quote(time()), 
			$this->connection->quote($this->config->calculateHash())
			);
		$result = $this->connection->query($sql);
		
		$row = $this->connection->fetchRow($result);
		return isset($row["session_data"]) ? $row["session_data"] : "";
	}
	
	function write($session_id, $session_data) {
		// insert OR update, read more here http://dev.mysql.com/doc/refman/4.1/en/insert-on-duplicate.html
		$sql = sprintf("INSERT INTO %s (session_id,hash,session_data,session_expire) VALUES (%s,%s,%s,%s)
			ON DUPLICATE KEY UPDATE session_data = %s, session_expire = %s", 
			$this->config->getTable(),
			$this->connection->quote($session_id),
			$this->connection->quote($this->config->calculateHash()),
			$this->connection->quote($session_data),
			$this->connection->quote(time() + $this->config->getSessionLifetime()),
			$this->connection->quote($session_data),
			$this->connection->quote(time() + $this->config->getSessionLifetime())
			);
		
		$result = $this->connection->query($sql);
		
		return $result ? true : false;
	}
}


src/Connection/PdoConnection.php
<?php
namespace steamegg\Slim\SessionDatabase\Connection;

abstract class PdoConnection implements IConnection {
	protected $pdo;
	
	abstract function getLock($name, $timeout);
	abstract function releaseLock($name);
	
	function __construct(\PDO $pdo){
		$this->pdo = $pdo;
	}
	
	function affectedRows(){
		return $this->pdo->affected_rows;
	}
	
	function error(){
		return $this->pdo->error;
	}
	
	/**
	 * @return \PDOStatement
	 */
	function query($query){
		return $this->pdo->query($query);
	}
	
	function quote($string){
		return $this->pdo->quote($string);
	}
	
	/**
	 * @param \PDOStatement
	 */
	function fetchRow($statement){
		return $statement->fetch(\PDO::FETCH_ASSOC);
	}
	
	function ping(){
		//PDO has no method like mysqli_ping()
		$statement = $this->pdo->query("SELECT 1");
		return $statement ? TRUE : FALSE;
	}
}


src/Connection/MysqliConnection.php
<?php
namespace steamegg\Slim\SessionDatabase\Connection;

class MysqliConnection implements IConnection {
	protected $db;
	
	function __construct(\mysqli $db){
		$this->db = $db;
	}
	
	function affectedRows(){
		return $this->db->affected_rows;
	}
	
	function error(){
		return $this->db->error;
	}
	
	function query($query){
		return $this->db->query($query);
	}
	
	function quote($string){
		return sprintf("'%s'", $this->escape($string));
	}
	
	protected function escape($string){
		return $this->db->real_escape_string($string);
	}
	
	function getLock($name, $timeout){
		$sql = sprintf("SELECT GET_LOCK('%s', '%s')", 
			$this->escape($name),
			$this->escape($timeout));
		
		$result = $this->query($sql);
		
		if(!$result)
			return FALSE;
		
		if(mysqli_num_rows($result) != 1)
			return FALSE;
		
		$row = mysqli_fetch_array($result);
		if($row[0] !== "1")
			return FALSE;
		
		return TRUE;
	}
	
	function releaseLock($name){
		$sql = sprintf("SELECT RELEASE_LOCK('%s')", $this->escape($name));
		$this->query($sql);
	}
	
	function fetchRow($result){
		return mysqli_fetch_assoc($result);
	}
	
	function ping(){
		return $this->db->ping();
	}
}


src/SessionDbHandler.php
<?php
namespace steamegg\Slim\SessionDatabase;

use steamegg\Slim\SessionDatabase\Connection\IConnection;

/**
 *  A PHP library acting as a drop-in replacement for PHP's default session handler, but instead of storing session data
 *  in flat files it stores them in a MySQL database, providing better performance as well as better security and
 *  protection against session fixation and session hijacking.
 *
 *  Read more {@link https://github.com/stefangabos/Zebra_Session/ here}
 *
 *  @author	 Stefan Gabos <contact@stefangabos.ro>
 *  @version	2.1.8 (last revision: May 20, 2017)
 *  @copyright  (c) 2006 - 2017 Stefan Gabos
 *  @license	http://www.gnu.org/licenses/lgpl-3.0.txt GNU LESSER GENERAL PUBLIC LICENSE
 *  @package	Zebra_Session
 *  
 *  ---------------------------------------------
 *  
 *  This is a fork of stefangabos/Zebra_Session.
 *  @author steamegg
 *  @package steamegg\Slim\SessionDatabase
 */

class SessionDbHandler implements \SessionHandlerInterface{

	/**
	 * @var IConnection
	 */
	private $connection;
	
	/**
	 * @var SessionConfig
	 */
	private $config;
	
	/**
	 * the maximum amount of time (in seconds) for which a process can lock the session
	 * @var int
	 */
	private $lock_timeout;
	
	/**
	 * @var string
	 */
	private $lockName;
	
	public function __construct(IConnection &$connection, SessionConfig $config, $lock_timeout = 60) {
		
		if(!$connection->ping())
			trigger_error("Connection ping failed", E_USER_ERROR);

		$this->connection = $connection;
		$this->config = $config;
		$this->lock_timeout = $lock_timeout;
		
		$this->setIni($this->config);

		session_set_save_handler(
			array(&$this, 'open'),
			array(&$this, 'close'),
			array(&$this, 'read'),
			array(&$this, 'write'),
			array(&$this, 'destroy'),
			array(&$this, 'gc')
		);

		session_start();
	}
	
	protected function setIni(SessionConfig $config){
		ini_set('session.gc_maxlifetime', $config->getSessionLifetime());
		ini_set('session.gc_probability', $config->getGcProbability());
		ini_set('session.gc_divisor', $config->getGcDivisor());
		
		// make sure session cookies never expire so that session lifetime
		// will depend only on the value of $session_lifetime
		ini_set('session.cookie_lifetime', 0);
		
		// tell the browser not to expose the cookie to client side scripting
		// this makes it harder for an attacker to hijack the session ID
		ini_set('session.cookie_httponly', 1);
		
		// make sure that PHP only uses cookies for sessions and disallow session ID passing as a GET parameter
		ini_set('session.use_only_cookies', 1);
	}

	function close() {
		if($this->connection->releaseLock($this->lockName))
			return TRUE;
	}

	function destroy($session_id) {
		$sql = sprintf("DELETE FROM %s WHERE session_id = %s", 
			$this->config->getTable(), 
			$this->connection->quote($session_id));
		$this->connection->query($sql);

		return $this->connection->affectedRows() >= 0 ? TRUE : FALSE;
	}

	function gc($maxlifetime) {
		$sql = sprintf("DELETE FROM %s WHERE session_expire < %s", 
			$this->config->getTable(), 
			$this->connection->quote(time()));
		$this->connection->query($sql);
	}

	function open($save_path, $name){
		return TRUE;
	}

	function read($session_id) {
		$this->lockName = sprintf("session_%s", $session_id);

		if( !$this->connection->getLock($this->lockName, $this->lock_timeout) )
			die("Could not obtain session lock");

		$sql = sprintf("SELECT session_data FROM %s WHERE session_id = %s AND session_expire > %s AND hash = %s LIMIT 1", 
			$this->config->getTable(), 
			$this->connection->quote($session_id), 
			$this->connection->quote(time()), 
			$this->connection->quote($this->config->calculateHash())
			);
		$result = $this->connection->query($sql);
		
		$row = $this->connection->fetchRow($result);
		return isset($row["session_data"]) ? $row["session_data"] : "";
	}
	
	function write($session_id, $session_data) {
		// insert OR update, read more here http://dev.mysql.com/doc/refman/4.1/en/insert-on-duplicate.html
		$sql = sprintf("INSERT INTO %s (session_id,hash,session_data,session_expire) VALUES (%s,%s,%s,%s)
			ON DUPLICATE KEY UPDATE session_data = %s, session_expire = %s", 
			$this->config->getTable(),
			$this->connection->quote($session_id),
			$this->connection->quote($this->config->calculateHash()),
			$this->connection->quote($session_data),
			$this->connection->quote(time() + $this->config->getSessionLifetime()),
			$this->connection->quote($session_data),
			$this->connection->quote(time() + $this->config->getSessionLifetime())
			);
		
		$result = $this->connection->query($sql);
		
		return $result ? true : false;
	}
}
